### PR TITLE
vc/qemu: add mutex to qmp monitor channel in qmpSetup()

### PR DESCRIPTION
**Pull request by YvesChan**
Solve possible race condition in qmpSetup()

Fixes: #2139

Signed-off-by: Yves Chan <shanks.cyp@gmail.com>

### CODE REVIEW

**Comment by devimc** [comment]
/test

thanks @YvesChan

**Comment by YvesChan** [comment]
> Thanks @YvesChan!
> 
> Is there any way we can add a test (either a unit test here, or a test in https://github.com/kata-containers/tests) to give us some confidence the race has been fixed?

As I mentioned above, now the containerd-shim-kata-v2 process will hold the qmp.socket all the time, so there's no chance to reproduce the race condition in current runtime version. But it's not very convenient for user to debug the qemu issue if they cannot access qmp.socket since it's held by shim. So it's just better to make qmpSetup() more robust. What if we change the qmp connection to ephemeral(free after used) in the future? who knows :)

File: virtcontainers/qemu.go [hunk 1]
```diff
@@ -921,6 +923,9 @@ func (q *qemu) togglePauseSandbox(pause bool) error {
 }
 
 func (q *qemu) qmpSetup() error {
+	q.qmpMonitorCh.Lock()
+	defer q.qmpMonitorCh.Unlock()
+
```
Comment (bergwolf) [inline virtcontainers/qemu.go]: If we get parallel qmp setup, is it possible to get parallel qmp shutdown as well?

Comment (bergwolf) [reply virtcontainers/qemu.go]: ping. The commit message says "Solve possible race condition in qmpSetup() and qmpShutdown()" but the shutdown part seems missing?

Comment (YvesChan) [reply virtcontainers/qemu.go]: odd... I have already pushed that commit. Anyway, I do it again.  Thanks :)

**Comment by jodh-intel** [comment]
This is now ready to land but I'll let @bergwolf click the lovely green button as an ack to the comment about shutdown.

**Pull request merged**